lib/client/ipam_test.go

```diff
@@ -0,0 +1,300 @@
+		Entry("Assign 1 IPv4 from a configured pool", net.IP{}, true, "pool1", net.IP{}, 66, []cnet.IP{}, nil),
+
+		// Test 5: Assign 1 IPv4 address with AssignIP then try to release 2 IPs.
+		// - Assign should not return an error.
+		// - ReleaseIPs should return a slice with one (unallocatedIPs) and an error???
+		Entry("Assign 1 IPv4 from a configured pool", net.IP{}, true, "pool1", net.ParseIP("192.168.1.0"), 0, []cnet.IP{}, nil),
+		Entry("Assign 1 IPv4 from a configured pool", net.ParseIP("192.168.1.1"), false, "pool1", net.IP{}, 0, []cnet.IP{cnet.IP{net.ParseIP("192.168.1.1")}}, nil),
+	)
+
+})
+
+// comment bro!
+// Using named returns to be clear which return value is which.
+func testIPAMReleaseIPs(inIP net.IP, pool string, cleanEnv bool, assignIP net.IP, autoAssignNumIPv4 int) (unallocatedIPs []cnet.IP, outErr error) {
+
+	inIPs := []cnet.IP{cnet.IP{inIP}}
+	if cleanEnv {
+		setupEnv(pool)
+	}
+	ic := setupIPMAClient(cleanEnv)
+
+	if len(assignIP) != 0 {
+		err := ic.AssignIP(client.AssignIPArgs{
+			IP: cnet.IP{assignIP},
+		})
+		if err != nil {
+			Fail(fmt.Sprintf("Error assigning IP %s", assignIP))
+		}
+
+		// Re-initialize it to an empty slice to flush out any IP if passed in my mistake.
+		inIPs = []cnet.IP{}
+
+		inIPs = append(inIPs, cnet.IP{assignIP})
+
+	}
+
+	if autoAssignNumIPv4 != 0 {
+		assignedIPv4, _, _ := ic.AutoAssign(client.AutoAssignArgs{
+			Num4: autoAssignNumIPv4,
+		})
+		inIPs = assignedIPv4
+	}
+
+	unallocatedIPs, outErr = ic.ReleaseIPs(inIPs)
+	if outErr != nil {
+		log.Println("######################: ", outErr)
+	}
+	return unallocatedIPs, outErr
+}
+
+// testIPAMAssignIP takes an IPv4 or IPv6 IP with a hostname and pool name and calls AssignIP.
+// When cleanEnv is passed it passes it along to wipe clean etcd and reset IPAM config.
+func testIPAMAssignIP(inIP net.IP, host, pool string, cleanEnv bool) error {
+	args := client.AssignIPArgs{
+		IP:       cnet.IP{inIP},
+		Hostname: host,
+	}
+	if cleanEnv {
+		setupEnv(pool)
+	}
+	ic := setupIPMAClient(cleanEnv)
+	outErr := ic.AssignIP(args)
+
+	if outErr != nil {
+		log.Println(outErr)
+	}
+	return outErr
+}
+
+// testIPAMAutoAssign takes number of requested IPv4 and IPv6, and hostname, and setus up/cleans up client and etcd,
+// then it calls AutoAssign (function under test) and returns the number of returned IPv4 and IPv6 addresses and returned error.
+func testIPAMAutoAssign(inv4, inv6 int, host string, cleanEnv bool, pool string) (int, int, error) {
+
+	args := client.AutoAssignArgs{
+		Num4:     inv4,
+		Num6:     inv6,
+		Hostname: host,
+	}
+
+	if cleanEnv {
+		setupEnv(pool)
+	}
+	ic := setupIPMAClient(cleanEnv)
+	v4, v6, outErr := ic.AutoAssign(args)
+
+	if outErr != nil {
+		log.Println(outErr)
+	}
+
+	return len(v4), len(v6), outErr
+}
+
+// setupEnv cleans up etcd if cleanEnv flag is passed and then creates IP pool based on the pool name passed to it.
+func setupEnv(pool string) {
+
+	etcdArgs := strings.Fields("rm /calico --recursive")
```

Comment:
nit: any reason not to use []string{"rm", "/calico", "--recursive"}  seems a bit convoluted to create a string and then split it.